### DISCUSSION TITLE
Creating a magit plugin

### DISCUSSION THREAD

**Discussion by david-greene-cb** [magit/magit#4709] [Support] [answered]
I would like to add some top-level fuctionality to magit, a new top-level command that will invoke a third-party git plugin.  I've looked through the magit and transient manuals and while it tells me how I might go about setting up the command menus, I can't find anything that clearly states how I plug that into magit itself.  Is there a hook I can use to add top-level commands?  Or some function to call that registers top-level commands?

Also, while I don't need this right now, it would also be great to know if there's a way to plug in additional infix/suffix commands to existing transients without directly modifying the code that sets up the transient.

**Comment by kyleam** [comment, marked as answer]
>  [..] how I plug that into magit itself. Is there a hook I can use to add top-level commands? Or some function to call that registers top-level commands?

Here's an example from magit-annex:

```elisp
;;;###autoload
(eval-after-load 'magit
  '(progn
     (define-key magit-mode-map "@" 'magit-annex-dispatch)
     (transient-append-suffix 'magit-dispatch '(0 -1 -1)
       '("@" "Annex" magit-annex-dispatch))))
```

https://github.com/magit/magit-annex/blob/efe484644666c6b7c544b0fb7b87e30703fa9425/magit-annex.el#L275-L280


> Also, while I don't need this right now, it would also be great to know if there's a way to plug in additional infix/suffix commands to existing transients without directly modifying the code that sets up the transient.

https://magit.vc/manual/transient/Modifying-Existing-Transients.html

However, for both of these, my recommendation is to not touch people's keymaps/transients.  Instead provide example snippets that your users can add to their config as they see fit.  This is what magit-imerge [now does](https://github.com/magit/magit-imerge/commit/bd548da8b0c982c346c5c3d11641b9602415ab74), and that's probably what magit-annex and magit-tbdiff _should_ have done.

(I'm curious what @tarsius's current opinion is, though.)

**Reply by tarsius** [reply]
(Also see https://github.com/magit/magit/wiki/Plugin-Dispatch-Key-Registry.)

> (I'm curious what @tarsius's current opinion is, though.)

I would keep the `magit-annex` and `magit-tbdiff` bindings for now. New extensions that we create should do it like `magit-imerge`, but there might be exceptions. `forge` started out unconditionally adding keymap and transient bindings, but now it only does so if `forge-add-default-bindings` is non-nil, which it is by default. That approach is a compromise, which I had to add because unconditionally adding the bindings posed an issue for those who maintain the `evil` bindings. We might have to do something similar for our other extensions, but I would wait until someone actually reports, that adding the bindings unconditionally, complicates things for them.

The same applies for third-party extensions, but here I lean more strongly towards not adding the bindings by default. Of course it is up to the author, at least until they want to add the package to Melpa. At which point they will likely be asked to not add the bindings by default.

If `magit` needs a new binding and the binding we choose is already used by an extension, then we should use it anyway.
